Add DgElementArrayMember

nilsdeppe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add DgElementArrayMember

kidder commented on src/Parallel/ArrayCollection/DgElementArrayMember.hpp:
```diff
@@ -0,0 +1,535 @@
+      tmpl::list<typename PhaseDepActionListsPack::action_list...>>;
+  /// The metavariables class passed to the Algorithm
+  using metavariables = Metavariables;
+  /// List off all the Tags that can be received into the Inbox
```
off -> of

kidder reviewed (changes requested):
nolint the clang-tidy error?

a lot of this code looks like it is copied from a DistributedObject.   Should these functions be factored out?


you can squash any changes

nilsdeppe commented:
1. added nolint for one. I don't know what to do about the other. The code seems to compile with clang so I don't know why clang-tidy is having issues.
2. Probably in the long run. I didn't mostly because we'll also need to test switching to a group instead of nodegroup to get it working with Kokkos, so just a lot of imminent changes that I don't know what they will require.
3. I've rebased and squashed. Thanks for the review!
